src/Backups/BackupIO_AzureBlobStorage.cpp

```diff
@@ -0,0 +1,365 @@
+            }
+
+
+            copyAzureBlobStorageFile(
+                client,
+                dest_client,
+                configuration.container,
+                fs::path(configuration.blob_path) / path_in_backup,
+                0,
+                file_size,
+                /* dest_bucket= */ blob_path[1],
+                /* dest_key= */ blob_path[0],
+                settings,
+                read_settings,
+                object_attributes,
+                threadPoolCallbackRunner<void>(getBackupsIOThreadPool().get(), "BackupReaderAzureBlobStorage"),
+                /* for_disk_azure_blob_storage= */ true);
+
+            return file_size;
+        };
+
+        destination_disk->writeFileUsingBlobWritingFunction(destination_path, write_mode, write_blob_function);
+        return; /// copied!
+    }
+
+    /// Fallback to copy through buffers.
+    BackupReaderDefault::copyFileToDisk(path_in_backup, file_size, encrypted_in_backup, destination_disk, destination_path, write_mode);
+}
+
+
+BackupWriterAzureBlobStorage::BackupWriterAzureBlobStorage(
+    StorageAzureBlob::Configuration configuration_,
+    const ReadSettings & read_settings_,
+    const WriteSettings & write_settings_,
+    const ContextPtr & context_)
+    : BackupWriterDefault(read_settings_, write_settings_, &Poco::Logger::get("BackupWriterAzureBlobStorage"))
+    , data_source_description{DataSourceType::AzureBlobStorage,configuration_.container, false, false}
+    , configuration(configuration_)
+{
+    client = StorageAzureBlob::createClient(configuration, /* is_read_only */ false);
+    settings = StorageAzureBlob::createSettingsAsSharedPtr(context_);
+    auto settings_as_unique_ptr = StorageAzureBlob::createSettings(context_);
+    object_storage = std::make_unique<AzureObjectStorage>("BackupWriterAzureBlobStorage",
+                                                          std::make_unique<Azure::Storage::Blobs::BlobContainerClient>(*client.get()),
+                                                                  std::move(settings_as_unique_ptr));
+}
+
+void BackupWriterAzureBlobStorage::copyFileFromDisk(const String & path_in_backup, DiskPtr src_disk, const String & src_path,
+                                      bool copy_encrypted, UInt64 start_pos, UInt64 length)
+{
+    /// Use the native copy as a more optimal way to copy a file from AzureBlobStorage to AzureBlobStorage if it's possible.
+    auto source_data_source_description = src_disk->getDataSourceDescription();
+    if (source_data_source_description.sameKind(data_source_description) && (source_data_source_description.is_encrypted == copy_encrypted))
+    {
+        /// getBlobPath() can return more than 3 elements if the file is stored as multiple objects in AzureBlobStorage bucket.
+        /// In this case we can't use the native copy.
+        if (auto blob_path = src_disk->getBlobPath(src_path); blob_path.size() == 2)
```

Comment:
you are right, will fix it. 
Any reason why getObjectNamespace is implemented to return bucket only for s3. For others we return empty. Just considering if we should change it to return container for AzureBlobStorage.

Reply:
We can override `getObjectNamespace()` for `AzureObjectStorage` of course, but what are you going to return with it? 

You can't return `std::shared_ptr<Azure::Storage::Blobs::BlobContainerClient>` via `getObjectNamespace()` because `getObjectNamespace()` must return a string.